shell/platform/darwin/ios/framework/Source/FlutterObservatoryPublisher.mm

```diff
@@ -153,24 +178,67 @@ - (void)netService:(NSNetService*)sender didNotPublish:(NSDictionary*)errorDict
                     "network settings and relaunch the application.";
 }
 
-#if TARGET_IPHONE_SIMULATOR
-static void DNSSD_API registrationCallback(DNSServiceRef sdRef,
-                                           DNSServiceFlags flags,
-                                           DNSServiceErrorType errorCode,
-                                           const char* name,
-                                           const char* regType,
-                                           const char* domain,
-                                           void* context) {
-  if (errorCode == kDNSServiceErr_NoError) {
-    FML_DLOG(INFO) << "FlutterObservatoryPublisher is ready!";
+@end
+
+@implementation FlutterObservatoryPublisher {
+  flutter::DartServiceIsolate::CallbackHandle _callbackHandle;
+  std::unique_ptr<fml::WeakPtrFactory<FlutterObservatoryPublisher>> _weakFactory;
+}
+
+- (NSURL*)url {
+  return [_delegate.get().url copy];
+}
+
+- (instancetype)init {
+  self = [super init];
+  NSAssert(self, @"Super must not return null on init.");
+
+  if (@available(iOS 9.3, *)) {
+    _delegate.reset([[ObservatoryDNSServiceDelegate alloc] initWithOwner:self]);
   } else {
-    FML_LOG(ERROR) << "Could not register as server for FlutterObservatoryPublisher. Check your "
-                      "network settings and relaunch the application.";
+    _delegate.reset([[ObservatoryNSNetServiceDelegate alloc] initWithOwner:self]);
   }
+  _weakFactory = std::make_unique<fml::WeakPtrFactory<FlutterObservatoryPublisher>>(self);
+
+  fml::MessageLoop::EnsureInitializedForCurrentThread();
+
+  _callbackHandle = flutter::DartServiceIsolate::AddServerStatusCallback(
+      [weak = _weakFactory->GetWeakPtr(),
+       runner = fml::MessageLoop::GetCurrent().GetTaskRunner()](const std::string& uri) {
+        if (!uri.empty()) {
+          runner->PostTask([weak, uri]() {
```

Comment:
Make a strong pointer in the block so it doesn't dealloc between the null check and the usage

Reply:
This is the pattern we use all over the place for fml::WeakPtrs - I think it's ok. We can't really get a "strong" reference because we're weakly holding it.